### PR TITLE
Disallow resetting attributes

### PR DESCRIPTION
**Pull request by deathowl**
**What:**
Disallow resetting attributes

**Why:**

#156 

**How:**
Make attributes set once

**Risks:**

**Checklist**:

- [X] Added tests, if you've added code that should be tested
- [X] Updated the documentation, if you've changed APIs
- [X] Ensured the test suite passes
- [X] Made sure your code lints
- [X] Completed the Contributor License Agreement ("CLA")

### CODE REVIEW

**Comment by deathowl** [comment]
Some Strcitmock related cases still fail in the mock_callable suite. As discussed with @fornellas  he will help in those.

**Comment by fornellas** [comment]
LGTM. I'm gonna be a bit busy in the following days, but after, I can give a hand with updating the tests.

File: testslide/__init__.py [hunk 1]
```diff
@@ -115,6 +115,16 @@ def _all_methods(self):
     def _all_memoizable_attributes(self):
         return self._context.all_context_data_memoizable_attributes
 
+    def __setattr__(self, name, value):
+        if self.__dict__.get(name) and self.__dict__[name] != value:
+            raise AttributeError(
+                f"Attribute: {name} is already set.\
+        Resetting attributes in sub-contexts is not allowed, as it breaks composition.\
+            Use memoize and composition instead"
+            )
```
Comment (fornellas) [inline testslide/__init__.py:124]: This is a non-trivial thing to understand for the average users. I suggest giving a more thorough and detailed message here.

PS: I think the white spaces from code formatting were part of the string.

```
                f"Attribute {repr(name)} is already set.\n"
                "Changing the value of attributes after they have been set "
                "can lead to unexpected test results. Eg: when a sub context "
                "resets an attribute after a parent context has set and used "
                "it, they will have different objects for the same attribute.\n"
                "You can safely override attributes from parent contexts by "
                "using @context.before, @context.memoize_before or "
                "@context.function, so the inner-most definition is used."
```

File: tests/mock_callable_testslide.py [hunk 2]
```diff
@@ -605,15 +605,15 @@ def to_raise_exception(context):
 
             @context.shared_context
             def integration(context):
-                @context.before
-                def catch_callable_target_exceptions(self):
-                    original_callable_target = self.callable_target
+                @context.memoize_before
```
Comment (fornellas) [inline tests/mock_callable_testslide.py:608]: Oh, this was tricky... I already regret having written it as it is, lol.

File: tests/mock_callable_testslide.py [hunk 3]
```diff
@@ -1111,6 +1111,10 @@ def mock_callable_dsl(self):
             def callable_target(self):
                 return getattr(self.real_target, self.callable_arg)
 
+            @context.memoize_before
+            def _original_target(self):
+                return getattr(self.real_target, self.callable_arg)
+
```
Comment (fornellas) [inline tests/mock_callable_testslide.py:1117]: Nit: I'd move this to line 607-608 so it is defined before it is referred for easier reading.

**Review by fornellas — APPROVED** [review]
LGTM.

I was thinking we could add a doc section under this one
https://testslide.readthedocs.io/en/master/testslide_dsl/context_attributes_and_functions/index.html#attributes
giving an example of resetting attributes, and how parent / child context may see different values and this being an issue, hooking to the next section of memoized attributes as the solution.

**Pull request merged**